File: Computer Networks/Question 4/linked-list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def insert_at_position(self, data, position):
        new_node = Node(data)
        if position == 1:
            new_node.next = self.head
            self.head = new_node
            return
        temp = self.head
        for _ in range(position - 2):
            if not temp:
                print("Invalid position!")
                return
            temp = temp.next
        new_node.next = temp.next
        temp.next = new_node

    def delete_by_value(self, value):
        if not self.head:
            print("List is empty!")
            return
        if self.head.data == value:
            self.head = self.head.next
            return
        temp = self.head
        while temp.next and temp.next.data != value:
            temp = temp.next
        if temp.next:
            temp.next = temp.next.next
        else:
            print(f"Value {value} not found!")

    def delete_by_position(self, position):
        if not self.head:
            print("List is empty!")
            return
        if position == 1:
            self.head = self.head.next
            return
        temp = self.head
        for _ in range(position - 2):
            if not temp.next:
                print("Invalid position!")
                return
            temp = temp.next
        if temp.next:
            temp.next = temp.next.next

    def search(self, value):
        temp = self.head
        position = 1
        while temp:
            if temp.data == value:
                print(f"Value {value} found at position {position}.")
                return
            temp = temp.next
            position += 1
        print(f"Value {value} not found!")

    def display(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("NULL")

# Demonstration
if __name__ == "__main__":
    ll = LinkedList()
    ll.insert_at_beginning(10)
    ll.insert_at_end(20)
    ll.insert_at_position(15, 2)
    print("List after insertions:")
    ll.display()

    ll.delete_by_value(15)
    print("List after deleting value 15:")
    ll.display()

    ll.delete_by_position(2)
    print("List after deleting position 2:")
    ll.display()

    ll.search(10)
    ll.search(30)

File: Computer Networks/Question 4/linked-list.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* insertAtBeginning(struct Node* head, int data);
struct Node* insertAtEnd(struct Node* head, int data);
struct Node* insertAtPosition(struct Node* head, int data, int position);
struct Node* deleteByValue(struct Node* head, int value);
struct Node* deleteByPosition(struct Node* head, int position);
void search(struct Node* head, int value);
void display(struct Node* head);

int main() {
    struct Node* head = NULL;
    printf("Linked List Operations:\n");

    head = insertAtBeginning(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtPosition(head, 15, 2);

    printf("List after insertions: ");
    display(head);

    head = deleteByValue(head, 15);
    printf("List after deleting value 15: ");
    display(head);

    head = deleteByPosition(head, 2);
    printf("List after deleting position 2: ");
    display(head);

    search(head, 10);
    search(head, 30);

    return 0;
}

struct Node* insertAtBeginning(struct Node* head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = head;
    return newNode;
}

struct Node* insertAtEnd(struct Node* head, int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;

    if (head == NULL) {
        return newNode;
    }

    struct Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = newNode;
    return head;
}

struct Node* insertAtPosition(struct Node* head, int data, int position) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;

    if (position == 1) {
        newNode->next = head;
        return newNode;
    }

    struct Node* temp = head;
    for (int i = 1; i < position - 1 && temp != NULL; i++) {
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("Invalid position!\n");
        free(newNode);
        return head;
    }

    newNode->next = temp->next;
    temp->next = newNode;
    return head;
}

struct Node* deleteByValue(struct Node* head, int value) {
    if (head == NULL) {
        printf("List is empty!\n");
        return head;
    }

    if (head->data == value) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* temp = head;
    while (temp->next != NULL && temp->next->data != value) {
        temp = temp->next;
    }

    if (temp->next == NULL) {
        printf("Value %d not found!\n", value);
    } else {
        struct Node* toDelete = temp->next;
        temp->next = toDelete->next;
        free(toDelete);
    }
    return head;
}

struct Node* deleteByPosition(struct Node* head, int position) {
    if (head == NULL) {
        printf("List is empty!\n");
        return head;
    }

    if (position == 1) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* temp = head;
    for (int i = 1; i < position - 1 && temp != NULL; i++) {
        temp = temp->next;
    }

    if (temp == NULL || temp->next == NULL) {
        printf("Invalid position!\n");
        return head;
    }

    struct Node* toDelete = temp->next;
    temp->next = toDelete->next;
    free(toDelete);
    return head;
}

void search(struct Node* head, int value) {
    struct Node* temp = head;
    int position = 1;
    while (temp != NULL) {
        if (temp->data == value) {
            printf("Value %d found at position %d.\n", value, position);
            return;
        }
        temp = temp->next;
        position++;
    }
    printf("Value %d not found!\n", value);
}

void display(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}